sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferStore.scala

```diff
@@ -134,135 +213,74 @@ abstract class RapidsBufferStore(
    */
-  protected def createBuffer(buffer: RapidsBuffer, memoryBuffer: MemoryBuffer, stream: Cuda.Stream)
-  : RapidsBufferBase
+  protected def createBuffer(
+     buffer: RapidsBuffer,
+     memoryBuffer: MemoryBuffer,
+     stream: Cuda.Stream): RapidsBufferBase
 
   /** Update bookkeeping for a new buffer */
-  protected def addBuffer(buffer: RapidsBufferBase): Unit = synchronized {
+  protected def addBuffer(buffer: RapidsBufferBase): Unit = {
     buffers.add(buffer)
-    catalog.registerNewBuffer(buffer)
   }
 
   override def close(): Unit = {
     buffers.freeAll()
   }
 
-  private def trySpillAndFreeBuffer(stream: Cuda.Stream): Long = synchronized {
-    val bufferToSpill = buffers.nextSpillableBuffer()
-    if (bufferToSpill != null) {
-      spillAndFreeBuffer(bufferToSpill, stream)
-      bufferToSpill.size
-    } else {
-      0
-    }
+  def nextSpillable(): RapidsBuffer = {
+    buffers.nextSpillableBuffer()
   }
 
-  private def spillAndFreeBuffer(buffer: RapidsBufferBase, stream: Cuda.Stream): Unit = {
-    if (spillStore == null) {
-      throw new OutOfMemoryError("Requested to spill without a spill store")
-    }
-    // If we fail to get a reference then this buffer has since been freed and probably best
-    // to return back to the outer loop to see if enough has been freed.
-    if (buffer.addReference()) {
-      try {
-        if (catalog.isBufferSpilled(buffer.id, buffer.storageTier)) {
-          logDebug(s"Skipping spilling $buffer ${buffer.id} to ${spillStore.name} as it is " +
-              s"already stored in multiple tiers total mem=${buffers.getTotalBytes}")
-          catalog.removeBufferTier(buffer.id, buffer.storageTier)
-        } else {
-          logDebug(s"Spilling $buffer ${buffer.id} to ${spillStore.name} " +
-              s"total mem=${buffers.getTotalBytes}")
-          val spillCallback = buffer.getSpillCallback
-          spillCallback(buffer.storageTier, spillStore.tier, buffer.size)
-          spillStore.copyBuffer(buffer, buffer.getMemoryBuffer, stream)
-        }
-      } finally {
-        buffer.close()
+  /**
+   * Waits for `targetTotalSize` to be available in this store up to FREE_WAIT_TIMEOUT.
+   * As pending buffers are closed, `pendingFreeBytes` is updated, and so we may exit
+   * from this function earlier than the timeout.
+   * @param targetTotalSize - amount of bytes we are waiting to have available
+   */
+  def waitForPending(targetTotalSize: Long): Unit = {
+    logWarning(s"Cannot spill further, waiting for ${pendingFreeBytes.get} " +
+      " bytes of pending buffers to be released")
+    memoryFreedMonitor.synchronized {
+      val memNeeded = buffers.getTotalSpillableBytes - targetTotalSize
+      if (memNeeded > 0 && memNeeded <= pendingFreeBytes.get) {
+        // This could be a futile wait if the thread(s) holding the pending buffers
+        // open are here waiting for more memory.
+        memoryFreedMonitor.wait(RapidsBufferStore.FREE_WAIT_TIMEOUT)
```

Comment:
This is a bit confusing. The docs say that we will wait for `targetTotalSize` to be available before waking up, or timed out.  But this only waits for any buffer to be freed before waking up. Also the code is doing an if/not a while or anything like that. I think @jlowe wanted us to exit sooner, which is fine, but then we should update the docs to indicate what this is really doing.

Reply:
With this https://github.com/NVIDIA/spark-rapids/pull/7572/commits/bc2a2150bfa8be133a14e9ee5a465eb0fd9d10ab we don't have the waits anymore.